README.md
# nextbus_chrome_extension

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-12-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-12-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/nextbus_chrome_extension/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/nextbus_chrome_extension/compare/0.1.0...0.1.1


test/nextbus_chrome_extension/core_test.clj
(ns nextbus-chrome-extension.core-test
  (:require [clojure.test :refer :all]
            [nextbus-chrome-extension.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


public/popup.js
/**
 * Populates error message in red text on the screen
 * @param {string} e error txt
 */
function displayErr (e) {
  document.querySelector('#errortxt').textContent = e;
}

/**
 * Format time in the form of mins away
 * @param {obj} time prediction record returne by api
 */
function formatTime (time) {
  // 2017-12-14T19:03:00Z
  var t = moment().to(moment(time.arrival_time, 'YYYY-MM-DDTHH:mm:ss'));
  var str = t.match(/in (\d+ min)/);
  return (str && typeof str[1] !== 'undefined' ? str[1] : '?');
}

/**
 * When the pop up renders make an ajax call to the api to get bus times
 */
document.addEventListener('DOMContentLoaded', () => {
  console.log('onload' + Date());
  var request = new XMLHttpRequest();
  request.open('GET', 'http://104.131.105.4:8081/', true);
  request.onload = function () {
    if (request.status >= 200 && request.status < 400) {
      // Success!
      var data = JSON.parse(request.responseText);
      console.log(data);
      var times = data.map((t) => {
        return formatTime(t.with[0]);
      });
      var els = document.querySelectorAll('p[id*=time');
      console.log({els:els});
      for (var i = 0; i < els.length; ++i) {
        els[i].textContent = times[i];
        els[i].classList.remove('loading');
      }
      console.log(times);
    } else {
      displayErr('Server returned error');
    }
  };
  request.onerror = function () {
    displayErr('Failed to connect to server');
  };
  request.send();
});


public/app.js
// ___UTILS___
var Impure = {
  // getJSON: R.curry(function (callback, url) {
  //   var request = new XMLHttpRequest();
  //   request.open('GET', url, true);
  //   request.onload = function () {
  //     if (request.status >= 200 && request.status < 400) {
  //       return 
  //     } else {
  //       displayErr('Server returned error');
  //     }
  //   };
  //   request.onerror = function () {
  //     displayErr('Failed to connect to server');
  //   };
  //   request.send();
  // }),
  getJSON: function get(url) {
    // Return a new promise.
    return new Promise(function(resolve, reject) {
      // Do the usual XHR stuff
      var req = new XMLHttpRequest();
      req.open('GET', url);
      req.onload = function() {
        // This is called even on 404 etc
        // so check the status
        if (req.status == 200) {
          // Resolve the promise with the response text
          resolve(req.response);
        }
        else {
          // Otherwise reject with the status text
          // which will hopefully be a meaningful error
          reject(Error(req.statusText));
        }
      };
      // Handle network errors
      req.onerror = function() {
        reject(Error("Network Error"));
      };
      // Make the request
      req.send();
    });
  },
  setHtml: R.curry(function (sel, html) {
    $(sel).html(html);
  })
};

// var trace = R.curry(function (tag, x) {
//     console.log(tag, x);
//     return x;
// });

// ___END UTIL___
// var url = function (t) {
//   return 'http://api.flickr.com/services/feeds/photos_public.gne?tags=' +
//     t + '&format=json&jsoncallback=?';
// };

// var mediaUrl = R.compose(R.prop('m'), R.prop('media'));
// var srcs = R.compose(R.map(mediaUrl), R.prop('items'));
// var images = R.compose(R.map(img), srcs);
// var renderImages = R.compose(Impure.setHtml('body'), images);
// var app = R.compose(Impure.getJSON(renderImages), url);
// var app = R.compose(Impure.getJSON, renderImages, url);

// a => [b]
var parseJson = function(r){
  return JSON.parse(r);
};

// const f = () => console.log('it is true')
// const g = () => console.log('nope')
// const condition = true

// // separate condition
// function truthyEither(c) {
//   return c ? RF.Either.Right("right") : RF.Either.Left("LEFT")
// }
// console.log(truthyEither(true));

var trace = R.curry(function (tag, x) {
  console.log(tag, x);
  return x;
});
var formatTime = (str) => {
  // 2017-12-14T19:03:00Z
  // str = 'afsd';
  var ts = moment().to(moment(str, 'YYYY-MM-DDTHH:mm:ss'));
  return ts.match(/in (\d+ min)/) ? RF.Either.Right(ts) : RF.Either.Left("No time");
};

var first = (arr) => {
  return arr.length > 0 ? RF.Either.Right(arr[0]) : RF.Either.Left("Missing First Elment of array");
};
// var log = (tag, x) => {
//   console.log(x)
// }
var time = (a) => a[0].arrival_time;
var arrivalTime = R.compose(formatTime, trace('arrival_time'), R.map(R.prop('arrival_time')),trace('first'), first, trace('with'), R.prop('with'));
var arrivalTimes = R.compose(R.map(arrivalTime), parseJson);

var testData = [
  {
    with: [
      {
        arrival_time: '2018-12-14T19:03:00Z'
      }
    ]
  }
];
console.log({test: testData.map(arrivalTime)});

var app = function(){
  Impure.getJSON('http://104.131.105.4:8081/').then((r) => {
    var sum = arrivalTimes(r);
    console.log({sum: sum});
  }).catch((e) => {
    console.log({error: e});
    document.querySelector('#errortxt').textContent = e;
  });
};
document.addEventListener('DOMContentLoaded', () => {
  // app();
});


